std/stdio.d

```diff
@@ -466,6 +466,283 @@ Throws: `ErrnoException` if the file could not be opened.
         std.file.remove(deleteme);
     }
 
+    /// Flags for file open mode.
+    enum OpenMode
+    {
+        /// Open in read mode.
+        read = 1 << 0,
+        /**
+         * Open in write mode, dont' truncate.
+         * Create a file if `existingOnly` flag is not provided.
+         */
+        write = 1 << 1,
+        /**
+         * Open in write mode, truncate if exists.
+         * Create a file if `existingOnly` flag is not provided.
+         */
+        truncate = 1 << 2,
+        /**
+         * Open in write mode. Append to the end on writing.
+         * Create a file if `existingOnly` flag is not provided.
+         *
+         * Note that it has a special meaning on Posix:
+         * the file opened in append mode will have write operations
+         * happening at the end of the file regardless of manual seek position changing.
+         */
+        append = 1 << 3,
+        /**
+         * Open in write mode. Create file only if it does not exist, error otherwise.
+         * The check for existence and the file creation is an atomical operation.
+         * Use this flag when need to ensure that the file with such name did not exist.
+         */
+        createNew = 1 << 4,
+        /**
+         * Don't create file if it does not exist when opening in write mode.
+         * This flag is not necessary when opening a file in read-only mode.
+         * This flag can't be used together with `OpenMode.createNew` flag.
+         */
+        existingOnly = 1 << 5,
+    }
+
+    private static bool hasAnyWriteFlag(OpenMode openMode) @safe nothrow pure @nogc
+    {
+        return (openMode & (OpenMode.write | OpenMode.truncate | OpenMode.append | OpenMode.createNew)) != 0;
+    }
+
+    private static const(char)[4] modezFromOpenMode(OpenMode openMode) @trusted
+    {
+        const bool anyWrite = hasAnyWriteFlag(openMode);
+        char[4] modez;
+        size_t i = 0;
+        if (openMode & OpenMode.read) {
+            if (anyWrite) {
+                if (openMode & OpenMode.append) {
+                    modez[i++] = 'a';
```

Comment:
This whole block has a poor style. You could return string literals, taking advantage of the automatic null sentinel, eg:

```d
        if (openMode & OpenMode.read)
        {
            if (anyWrite)
            {
                if (openMode & OpenMode.append)
                    return "a+"; // or "a+\0\0" if necessary
                else if (openMode & OpenMode.existingOnly)
                    return "r+";
                else if (openMode & OpenMode.truncate)
                    return "rw";
                else
                   return "w+";
            }
            else return "r";
        }
 // etc....
```